integration/test/validateliveresources/helpers.go

```diff
@@ -0,0 +1,161 @@
+// +build liveinstallation
+
+package validateliveresources
+
+import (
+	"context"
+	"os"
+	"testing"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2019-07-01/compute"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	corev1alpha1 "github.com/giantswarm/apiextensions/v3/pkg/apis/core/v1alpha1"
+	"github.com/giantswarm/apiextensions/v3/pkg/apis/provider/v1alpha1"
+	releasev1alpha1 "github.com/giantswarm/apiextensions/v3/pkg/apis/release/v1alpha1"
+	securityv1alpha1 "github.com/giantswarm/apiextensions/v3/pkg/apis/security/v1alpha1"
+	"github.com/giantswarm/micrologger"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	capz "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+	capzexp "sigs.k8s.io/cluster-api-provider-azure/exp/api/v1alpha3"
+	capi "sigs.k8s.io/cluster-api/api/v1alpha3"
+	capiexp "sigs.k8s.io/cluster-api/exp/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+
+	"github.com/giantswarm/azure-admission-controller/integration/env"
+	"github.com/giantswarm/azure-admission-controller/internal/vmcapabilities"
+)
+
+func NewReadOnlyCtrlClient(t *testing.T) client.Client {
```

Comment:
Could we use the controller-runtime reader/cache thingy instead? That interface didn't have the writing methods if I remember correctly? So we wouldn't need our own implementation

Reply:
> That interface didn't have the writing methods if I remember correctly? So we wouldn't need our own implementation

True, I started with that one, then saw that we use `client.Client` in all structs in the project, so it's not possible to pass `client.Reader`. We would have to change all existing `client.Client` usage to `client.Reader` instead, which would probably be even better, as we would be sure that we are not modifying stuff accidentally.

Another caveat there is that we also already have `client.Reader` in some structs, which we use for cache purposes (since `client.Reader` is a subset interface of `client.Cache`).
So if we don't want to have two `client.Reader` fields in some places, we would have to also change `client.Reader` into `client.Cache` there, aaand this was the part where I stopped, since I was not fully sure why Tuomas used `client.Reader` instead of `client.Cache` in the first place, but one thing that I'm sure is that the tests are simpler with `client.Reader`.

So there you go :slightly_smiling_face: I took a bit shorter path (this little read-only wrapper struct) as you can see :smile: